core/ArchiveProcessor/RecordBuilder.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+namespace Piwik\ArchiveProcessor;
+
+use Piwik\ArchiveProcessor;
+use Piwik\Common;
+use Piwik\DataTable;
+
+/**
+ * Inherit from this class to define archiving logic for one or more records.
+ */
+abstract class RecordBuilder
+{
+    /**
+     * @var int
+     */
+    protected $maxRowsInTable;
+
+    /**
+     * @var int
+     */
+    protected $maxRowsInSubtable;
+
+    /**
+     * @var string|null
+     */
+    protected $columnToSortByBeforeTruncation;
+
+    /**
+     * @var int
+     */
+    protected $blobReportLimit;
+
+    /**
+     * @var array|null
+     */
+    protected $columnAggregationOps;
+
+    /**
+     * @param int|null $maxRowsInTable
+     * @param int|null $maxRowsInSubtable
+     * @param string|null $columnToSortByBeforeTruncation
+     * @param array|null $columnAggregationOps
+     */
+    public function __construct(?int $maxRowsInTable = null, ?int $maxRowsInSubtable = null,
+                                ?string $columnToSortByBeforeTruncation = null, ?array $columnAggregationOps = null)
+    {
+        $this->maxRowsInTable = $maxRowsInTable;
+        $this->maxRowsInSubtable = $maxRowsInSubtable;
+        $this->columnToSortByBeforeTruncation = $columnToSortByBeforeTruncation;
+        $this->columnAggregationOps = $columnAggregationOps;
+    }
+
+    public function isEnabled(ArchiveProcessor $archiveProcessor): bool
+    {
+        return true;
+    }
+
+    public function build(ArchiveProcessor $archiveProcessor): void
+    {
+        if (!$this->isEnabled($archiveProcessor)) {
+            return;
+        }
+
+        $recordsBuilt = $this->getRecordMetadata($archiveProcessor);
+
+        $recordMetadataByName = [];
+        foreach ($recordsBuilt as $recordMetadata) {
+            $recordMetadataByName[$recordMetadata->getName()] = $recordMetadata;
+        }
+
+        $numericRecords = [];
+
+        $records = $this->aggregate($archiveProcessor);
+        foreach ($records as $recordName => $recordValue) {
+            if (empty($recordMetadataByName[$recordName])) {
```

Comment:
`array_key_exists` or `isset` might be more accurate here even though it behaves the same as empty in terms of checking for an unset key. In terms of code understanding it helps better with the intentions, empty might be checking for null or false or 0 which is not the case here.


```suggestion
            if (!array_key_exists($recordMetadataByName[$recordName])) {
```

Reply:
If for some reason a plugin (third party or no) returns 0, null or false here, we don't want the code to outright fail.